In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [3]:
# import oder.csv
df_prods = pd.read_csv(r'C:\Users\User\Documents\Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [5]:
path = r'C:\Users\User\Documents\Instacart Basket Analysis'

In [7]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [9]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [11]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [15]:
# finding missing value in product df
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [17]:
# to view missing values
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [19]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [21]:
df_prods.shape

(49693, 5)

In [23]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
df_prods_clean.shape

(49677, 5)

In [27]:
# finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [29]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [31]:
df_prods_clean.shape

(49677, 5)

In [35]:
# drop duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [37]:
df_prods_clean_no_dups.shape

(49672, 5)

In [99]:
# Export product clean data frame
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index = False)

In [41]:
# step 2 investigate order dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

 # findings
 As the count of days_since_prior_order is lower than the other columns, it seems there are some missing values in this column.
 Order no. has min value as 1 and max value as 100. The days of the week are grom 0 to 6 and hence there seems to be no outliers here. Similarly, the order hour of the day values range from 1-23 which indicates that there are no outliers there as well. 
 

In [45]:
# check mixed data type in df ords data frame
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [47]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

step 4: There are no mixed data type in order data frame

In [51]:
# step 5 : check for missing value in df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in the 'days_since_prior_order' column. 

In [57]:
# create a subset containing only missing values:
df_ords_nan = df_ords[df_ords['days_since_prior_order']. isnull() == True]

In [59]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [61]:
# check the frequeerncy of order number1rder_number
df_ords_nan['order_number'].value_counts()

order_number
1    206209
Name: count, dtype: int64

The result shows frequency of missing values 206209 is the same as frequency of order number 1.

In [75]:
# create a new column to flag the reason for missing values
df_ords['first_order'] = df_ords['days_since_prior_order'].isna()

In [77]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  First Order  first_order  
0                        8                     NaN         True         True  
1                        7                    15.0        False        False  
2                       12                    21.0        False        False  
3                        7                    29.0        False        False  
4                       15                    28.0        False        False  
...                    ...                     ...          ...          ...  
3421078                 18                    29.0        False        False  
3421079                 10                    30.0        False        False  
3421080                 12                    18.0        False        False  
3421081                 12                     7.0        False        False  
3421082                 14                    30.0        False        False  

[3421083 rows x 9 columns]

In [79]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
First Order                    0
first_order                    0
dtype: int64

In [81]:
df_ords = df_ords.drop(columns = ['First Order'])


In [83]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
first_order                    0
dtype: int64

In [85]:
df_ords['first_order'].value_counts()

first_order
False    3214874
True      206209
Name: count, dtype: int64

In [89]:
# step 7 : check duplicate values
df_ords_dups = df_ords[df_ords.duplicated()]

In [91]:
df_ords_dups.shape

(0, 8)

There is no missing value in df order dataframe.

In [97]:
# Export order data frame
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index = False)